src/Media.tsx

```diff
@@ -365,12 +368,13 @@ export function createMedia<
               const matchingMediaQueries = Object.keys(matches).filter(
                 key => matches[key]
               )
+
+              const MediaContextValue = getMediaContextValue(
+                intersection(matchingMediaQueries, onlyMatch)
+              )
+
               return (
-                <MediaContext.Provider
-                  value={{
-                    onlyMatch: intersection(matchingMediaQueries, onlyMatch),
-                  }}
-                >
+                <MediaContext.Provider value={MediaContextValue}>
```

Comment:
It's applying this value that actually fixes the re-render in the `Media` components